servatrice/src/serversocketinterface.cpp

```diff
@@ -1054,14 +1057,93 @@ Response::ResponseCode AbstractServerSocketInterface::cmdAccountPassword(const C
+{
+	qDebug() << "Received forgot password request from user: " << QString::fromStdString(cmd.user_name());
+
+	if (!servatrice->getEnableForgotPassword())
+		return Response::RespFunctionNotAllowed;
+
+	if (!sqlInterface->userExists(QString::fromStdString(cmd.user_name())))
+		return Response::RespFunctionNotAllowed;
+
+	if (sqlInterface->doesForgotPasswordExist(QString::fromStdString(cmd.user_name()))) {
+		Response_ForgotPasswordRequest *re = new Response_ForgotPasswordRequest;
+		re->set_challenge_email(false);
+		rc.setResponseExtension(re);
+		return Response::RespOk;
+	}
+
+	QString banReason; int banTimeRemaining;
+	if (sqlInterface->checkUserIsBanned(this->getAddress(), QString::fromStdString(cmd.user_name()), QString::fromStdString(cmd.clientid()), banReason, banTimeRemaining))
+		return Response::RespFunctionNotAllowed;
+
+	if (servatrice->getEnableForgotPasswordChallenge()) {
+		Response_ForgotPasswordRequest *re = new Response_ForgotPasswordRequest;
+		re->set_challenge_email(true);
+		rc.setResponseExtension(re);
+		return Response::RespOk;
+	}
+	else {
+		if (sqlInterface->addForgotPassword(QString::fromStdString(cmd.user_name()))) {
+			Response_ForgotPasswordRequest *re = new Response_ForgotPasswordRequest;
+			re->set_challenge_email(false);
+			rc.setResponseExtension(re);
+			return Response::RespOk;
+		}
+	}
+
+	return Response::RespFunctionNotAllowed;
+}
+
+Response::ResponseCode AbstractServerSocketInterface::cmdForgotPasswordReset(const Command_ForgotPasswordReset &cmd, ResponseContainer &rc)
+{
+	Q_UNUSED(rc);
+	qDebug() << "Received forgot password reset from user: " << QString::fromStdString(cmd.user_name());
+
+	if (!sqlInterface->doesForgotPasswordExist(QString::fromStdString(cmd.user_name())))
+		return Response::RespFunctionNotAllowed;
+
+	// if (!sqlInterface->validateUserToken(QString::fromStdString(cmd.token()), QString::fromStdString(cmd.user_name())))
+	//	return Response::RespFunctionNotAllowed;
+
+	if (!sqlInterface->validateTableColumnStringData("{prefix}_users","token", QString::fromStdString(cmd.user_name()),QString::fromStdString(cmd.token())))
+		return Response::RespFunctionNotAllowed;
+
+	if (sqlInterface->changeUserPassword(QString::fromStdString(cmd.user_name()), "", QString::fromStdString(cmd.new_password()), true)) {
+		sqlInterface->removeForgotPassword(QString::fromStdString(cmd.user_name()));
+		return Response::RespOk;
+	}
+
+	return Response::RespFunctionNotAllowed;
+}
+
+Response::ResponseCode AbstractServerSocketInterface::cmdForgotPasswordChallenge(const Command_ForgotPasswordChallenge &cmd, ResponseContainer &rc)
+{
+	Q_UNUSED(rc);
+	qDebug() << "Received forgot password challenge from user: " << QString::fromStdString(cmd.user_name());
+
+	if (sqlInterface->doesForgotPasswordExist(QString::fromStdString(cmd.user_name())))
+		return Response::RespOk;
+
+	//if (sqlInterface->validateUserEmail(QString::fromStdString(cmd.email()), QString::fromStdString(cmd.user_name())))
```

Comment:
Yup, this was commented out when I built the challenge questions.  I'll remove them.